src/python/pants/rules/core/filedeps.py

```diff
@@ -44,40 +62,45 @@ class Filedeps(Goal):
 
 @goal_rule
 async def file_deps(
-    console: Console,
-    options: FiledepsOptions,
-    build_root: BuildRoot,
-    transitive_hydrated_targets: TransitiveHydratedTargets,
+    console: Console, options: FiledepsOptions, build_root: BuildRoot, addresses: Addresses,
 ) -> Filedeps:
-    unique_rel_paths: Set[str] = set()
+    targets: Iterable[Target]
+    if options.values.transitive:
+        transitive_targets = await Get[TransitiveTargets](Addresses, addresses)
+        targets = transitive_targets.closure
+    else:
+        targets = await Get[Targets](Addresses, addresses)
+
     build_file_addresses = await MultiGet(
-        Get[BuildFileAddress](Address, hydrated_target.adaptor.address)
-        for hydrated_target in transitive_hydrated_targets.closure
+        Get[BuildFileAddress](Address, tgt.address) for tgt in targets
     )
-    for hydrated_target, build_file_address in zip(
-        transitive_hydrated_targets.closure, build_file_addresses
-    ):
-        unique_rel_paths.add(build_file_address.rel_path)
-        adaptor = hydrated_target.adaptor
-        if adaptor.has_sources():
-            sources_paths = (
-                adaptor.sources.snapshot.files
-                if not options.values.globs
-                else adaptor.sources.filespec["globs"]
+    unique_rel_paths = {bfa.rel_path for bfa in build_file_addresses}
+
+    if options.values.globs:
+        unique_rel_paths.update(
+            itertools.chain.from_iterable(tgt.get(Sources).filespec["globs"] for tgt in targets)
```

Comment:
FYI on how the Target API works: if the target type does not have the `Sources` field registered, then `tgt.get()` will create a default instance of the `Sources` field by calling `Sources(raw_value=None)`.